compiler/z/codegen/OMRLinkage.cpp

```diff
@@ -1585,15 +1583,34 @@ OMR::Z::Linkage::copyArgRegister(TR::Node * callNode, TR::Node * child, TR::Regi
 
    TR_Debug * debugObj = self()->cg()->getDebug();
    char * REG_PARAM = "LR=Reg_param";
-   if (!self()->cg()->canClobberNodesRegister(child, 0))
+   if (TR::Compiler->target.is32Bit() && self()->cg()->use64BitRegsOn32Bit() && child->getDataType() == TR::Int64 && !argRegister->getRegisterPair())
       {
-      if (!(TR::Compiler->target.is64Bit()) &&
-           ((child->getDataType() == TR::Int64 ) || (child->getDataType() == TR::Int64 )))
+      TR::Register * tempRegH = self()->cg()->allocateRegister();
+      TR::Register * tempRegL = self()->cg()->allocateRegister();
+
+      cursor = generateRSInstruction(self()->cg(), TR::InstOpCode::SRLG,
+          callNode, tempRegH, argRegister, 32);
+      if (debugObj)
          {
-         TR::Register * tempRegH=NULL;
-         TR::Register * tempRegL=NULL;
-         tempRegH = self()->cg()->allocateRegister();
-         tempRegL = self()->cg()->allocateRegister();
+         debugObj->addInstructionComment(toS390RSInstruction(cursor), REG_PARAM);
+         }
+
+      cursor = generateRRInstruction(self()->cg(), copyOpCode,
+        callNode, tempRegL, argRegister);
+      if (debugObj)
+         {
+         debugObj->addInstructionComment(toS390RRInstruction(cursor), REG_PARAM);
+         }
+
+      argRegister = self()->cg()->allocateConsecutiveRegisterPair(tempRegL, tempRegH);
+      self()->cg()->stopUsingRegister(argRegister);
+      }
+   else if (!self()->cg()->canClobberNodesRegister(child, 0))
+      {
+      if (TR::Compiler->target.is32Bit() && child->getDataType() == TR::Int64)
```

Comment:
In a previous discussion you mentioned that if we enable 64 bit registers on 32 bit, no long values should be passed in register pairs and we should add an assertion here since this would be dead code. Before, the assertion used here was `TR_ASSERT_FATAL(0, "Long values should not be passed in register pairs, they can be stored in a single GPR");`. However, this would cause failures when disabling these changes, Should I add an assertion such as `TR_ASSERT_FATAL(self()->cg()->use64BitRegsOn32Bit(), "Long values should not be passed in register pairs, they can be stored in a single GPR");` or do you think it is unnecessary?

Reply:
I don't think it hurts to have the assert added if that is the expectation, and AFAIK it should be. Please make sure this is tested when you add the assert back.

Reply:
In tests from the discussion https://github.com/eclipse/omr/pull/2916#discussion_r217520385 I had the assert when 64 bit regs were enabled and commented out the assert otherwise. This path should be well tested. 

57d3672